forge/src/executor/inspector/cheatcodes/expect.rs

```diff
@@ -0,0 +1,203 @@
+use super::Cheatcodes;
+use crate::abi::HEVMCalls;
+use bytes::Bytes;
+use ethers::{
+    abi::{AbiEncode, RawLog},
+    types::{Address, H256},
+};
+use once_cell::sync::Lazy;
+use revm::{return_ok, Database, EVMData, Interpreter, Return};
+use std::str::FromStr;
+
+/// For some cheatcodes we may internally change the status of the call, i.e. in `expectRevert`.
+/// Solidity will see a successful call and attempt to decode the return data. Therefore, we need
+/// to populate the return with dummy bytes so the decode doesn't fail.
+static DUMMY_CALL_OUTPUT: [u8; 320] = [0u8; 320];
+
+/// Same reasoning as [DUMMY_CALL_OUTPUT], but for creates.
+static DUMMY_CREATE_ADDRESS: Lazy<Address> =
+    Lazy::new(|| Address::from_str("0000000000000000000000000000000000000001").unwrap());
+
+#[derive(Clone, Debug, Default)]
+pub struct ExpectedRevert {
+    /// The expected data returned by the revert
+    pub reason: Bytes,
+    /// The depth at which the revert is expected
+    pub depth: u64,
+}
+
+fn expect_revert(state: &mut Cheatcodes, reason: Bytes, depth: u64) -> Result<Bytes, Bytes> {
+    if state.expected_revert.is_some() {
+        Err("You must call another function prior to expecting a second revert."
+            .to_string()
+            .encode()
+            .into())
+    } else {
+        state.expected_revert = Some(ExpectedRevert { reason, depth });
+        Ok(Bytes::new())
+    }
+}
+
+pub fn handle_expect_revert(
+    is_create: bool,
+    expected_revert: &Bytes,
+    status: Return,
+    retdata: Bytes,
+) -> Result<(Option<Address>, Bytes), Bytes> {
+    if matches!(status, return_ok!()) {
+        return Err("Call did not revert as expected".to_string().encode().into())
+    }
+
+    if retdata.is_empty() {
```

Comment:
Should this error only happen if expect_revert is not empty? Thinking about the expectRevert(bytes("")) case

Reply:
I think this is correct, it's a condensed form of

https://github.com/gakonst/foundry/blob/a00c1939becf3b40f1453e330d3d9507a756d610/evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs#L452-L468

In the case of `expectRevert(bytes(""))` I think there is still data (some `bytes` header)

Would make sense to have some extra eyes on it though

Reply:
This was a bug but it is fixed now